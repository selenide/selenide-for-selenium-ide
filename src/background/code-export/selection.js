import {codeExport as exporter} from '@seleniumhq/side-utils'

const emitters = {
  id: emitId,
  value: emitValue,
  label: emitLabel,
  index: emitIndex,
}

export function emit(location) {
  return exporter.emit.selection(location, emitters)
}

export default {
  emit,
}

function emitId(id) {
  return Promise.resolve(`$(By.cssSelector("#${id}")).click();`)
}

function emitValue(value) {
  return Promise.resolve(`selectOptionByValue("${value}");`)
}

function emitLabel(label) {
  return Promise.resolve(`selectOption("${label}");`)
}

function emitIndex(index) {
  return Promise.resolve(`selectOption(${index});`)
}
