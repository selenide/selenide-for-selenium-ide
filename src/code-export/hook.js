import {codeExport as exporter} from '@seleniumhq/side-utils';

const emitters = {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  declareDependencies,
  declareMethods: empty,
  declareVariables,
  inEachBegin: empty,
  inEachEnd: empty
};

function generate(hookName) {
  return new exporter.hook(emitters[hookName]());
}

export function generateHooks() {
  let result = {};
  Object.keys(emitters).forEach(hookName => {
    result[hookName] = generate(hookName);
  });
  return result;
}

function afterAll() {
  const params = {
    startingSyntax: {
      commands: [
        {level: 0, statement: '@AfterAll'},
        {level: 0, statement: 'public static void finalTearDown() {'}
      ]
    },
    endingSyntax: {
      commands: [{level: 0, statement: '}'}]
    },
    registrationLevel: 1
  };
  return params;
}

function afterEach() {
  const params = {
    startingSyntax: {
      commands: [
        {level: 0, statement: '@AfterEach'},
        {level: 0, statement: 'public void tearDown() {'}
      ]
    },
    endingSyntax: {
      commands: [{level: 0, statement: '}'}]
    }
  };
  return params;
}

function beforeAll() {
  return {
    startingSyntax: {
      commands: [
        {level: 0, statement: '@BeforeAll'},
        {level: 0, statement: 'public static void initialSetUp() {'}
      ]
    },
    endingSyntax: {
      commands: [{level: 0, statement: '}'}]
    },
    registrationLevel: 1
  };
}

function beforeEach() {
  return {
    startingSyntax: ({browserName, gridUrl} = {}) => ({
      commands: [
        {level: 0, statement: '@BeforeEach'},
        {
          level: 0,
          statement: `public void setUp() {`
        },
        {
          level: 1,
          statement: gridUrl
            ? `Configuration.browser = "${browserName ? browserName.toLowerCase() : 'chrome'}";\n    Configuration.remote = "${gridUrl}";`
            : `Configuration.browser = "${browserName ? browserName.toLowerCase() : 'chrome'}";`
        },
        {level: 1, statement: 'vars = new HashMap<>();'}
      ]
    }),
    endingSyntax: {
      commands: [{level: 0, statement: '}'}]
    }
  };
}

function declareDependencies() {
  const params = {
    startingSyntax: {
      commands: [
        {level: 0, statement: 'import org.junit.jupiter.api.Test;'},
        {level: 0, statement: 'import org.junit.jupiter.api.BeforeAll;'},
        {level: 0, statement: 'import org.junit.jupiter.api.BeforeEach;'},
        {level: 0, statement: 'import org.junit.jupiter.api.AfterAll;'},
        {level: 0, statement: 'import org.junit.jupiter.api.AfterEach;'},
        {level: 0, statement: 'import com.codeborne.selenide.Configuration;'},
        {level: 0, statement: 'import com.codeborne.selenide.WebDriverRunner;'},
        {level: 0, statement: 'import static com.codeborne.selenide.Selectors.*;'},
        {level: 0, statement: 'import static com.codeborne.selenide.Condition.*;'},
        {level: 0, statement: 'import static com.codeborne.selenide.Selenide.*;'},
        {level: 0, statement: 'import static org.junit.jupiter.api.Assertions.*;'},
        {level: 0, statement: 'import org.openqa.selenium.Dimension;'},
        {level: 0, statement: 'import org.openqa.selenium.Keys;'},
        {level: 0, statement: 'import java.util.*;'}
      ]
    }
  };
  return params;
}

function declareVariables() {
  const params = {
    startingSyntax: {
      commands: [
        {
          level: 0,
          statement: 'private Map<String, Object> vars;'
        }
      ]
    }
  };
  return params;
}

function empty() {
  return {};
}
