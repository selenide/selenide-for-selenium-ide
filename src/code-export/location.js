"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.emit = emit;
exports.default = void 0;

let _sideUtils = require("@seleniumhq/side-utils");

const emitters = {
  id: emitId,
  name: emitName,
  link: emitLink,
  linkText: emitLink,
  partialLinkText: emitPartialLinkText,
  css: emitCss,
  xpath: emitXpath
};

function emit(location) {
  return _sideUtils.codeExport.emit.location(location, emitters);
}

let _default = {
  emit
};
exports.default = _default;

function emitId(selector) {
  return Promise.resolve(`"#${selector}"`);
}

function emitName(selector) {
  return Promise.resolve(`byName("${selector}")`);
}

function emitLink(selector) {
  return Promise.resolve(`byLinkText("${selector}")`);
}

function emitPartialLinkText(selector) {
  return Promise.resolve(`byPartialLinkText("${selector}")`);
}

function emitCss(selector) {
  return Promise.resolve(`"${selector}"`);
}

function emitXpath(selector) {
  return Promise.resolve(`byXpath("${selector}")`);
}
