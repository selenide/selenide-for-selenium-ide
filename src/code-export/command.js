import {codeExport as exporter} from '@seleniumhq/side-utils';
import location from './location';
import selection from './selection';

export const emitters = {
  addSelection: emitSelect,
  answerOnNextPrompt: skip,
  assert: emitAssert,
  assertAlert: emitAssertAlert,
  assertChecked: emitVerifyChecked,
  assertConfirmation: emitAssertAlert,
  assertEditable: emitVerifyEditable,
  assertElementPresent: emitVerifyElementPresent,
  assertElementNotPresent: emitVerifyElementNotPresent,
  assertNotChecked: emitVerifyNotChecked,
  assertNotEditable: emitVerifyNotEditable,
  assertNotSelectedValue: emitVerifyNotSelectedValue,
  assertNotText: emitVerifyNotText,
  assertPrompt: emitAssertAlert,
  assertSelectedLabel: emitVerifySelectedLabel,
  assertSelectedValue: emitVerifySelectedValue,
  assertValue: emitVerifyValue,
  assertText: emitVerifyText,
  assertTitle: emitVerifyTitle,
  check: emitCheck,
  chooseCancelOnNextConfirmation: skip,
  chooseCancelOnNextPrompt: skip,
  chooseOkOnNextConfirmation: skip,
  click: emitClick,
  clickAt: emitClick,
  close: emitClose,
  debugger: skip,
  do: emitControlFlowDo,
  doubleClick: emitDoubleClick,
  doubleClickAt: emitDoubleClick,
  dragAndDropToObject: emitDragAndDrop,
  echo: emitEcho,
  editContent: emitEditContent,
  else: emitControlFlowElse,
  elseIf: emitControlFlowElseIf,
  end: emitControlFlowEnd,
  executeScript: emitExecuteScript,
  executeAsyncScript: emitExecuteAsyncScript,
  forEach: emitControlFlowForEach,
  if: emitControlFlowIf,
  mouseDown: emitMouseDown,
  mouseDownAt: emitMouseDown,
  mouseMove: emitMouseMove,
  mouseMoveAt: emitMouseMove,
  mouseOver: emitMouseMove,
  mouseOut: emitMouseOut,
  mouseUp: emitMouseUp,
  mouseUpAt: emitMouseUp,
  open: emitOpen,
  pause: emitPause,
  removeSelection: emitSelect,
  repeatIf: emitControlFlowRepeatIf,
  run: emitRun,
  runScript: emitRunScript,
  select: emitSelect,
  selectFrame: emitSelectFrame,
  selectWindow: emitSelectWindow,
  sendKeys: emitSendKeys,
  setSpeed: emitSetSpeed,
  setWindowSize: emitSetWindowSize,
  store: emitStore,
  storeAttribute: emitStoreAttribute,
  //storeJson: emitStoreJson,
  storeText: emitStoreText,
  storeTitle: emitStoreTitle,
  storeValue: emitStoreValue,
  storeWindowHandle: emitStoreWindowHandle,
  storeXpathCount: emitStoreXpathCount,
  submit: emitSubmit,
  times: emitControlFlowTimes,
  type: emitType,
  uncheck: emitUncheck,
  verify: emitAssert,
  verifyChecked: emitVerifyChecked,
  verifyEditable: emitVerifyEditable,
  verifyElementPresent: emitVerifyElementPresent,
  verifyElementNotPresent: emitVerifyElementNotPresent,
  verifyNotChecked: emitVerifyNotChecked,
  verifyNotEditable: emitVerifyNotEditable,
  verifyNotSelectedValue: emitVerifyNotSelectedValue,
  verifyNotText: emitVerifyNotText,
  verifySelectedLabel: emitVerifySelectedLabel,
  verifySelectedValue: emitVerifySelectedValue,
  verifyText: emitVerifyText,
  verifyTitle: emitVerifyTitle,
  verifyValue: emitVerifyValue,
  waitForElementEditable: emitWaitForElementEditable,
  waitForElementPresent: emitWaitForElementPresent,
  waitForElementVisible: emitWaitForElementVisible,
  waitForElementNotEditable: emitWaitForElementNotEditable,
  waitForElementNotPresent: emitWaitForElementNotPresent,
  waitForElementNotVisible: emitWaitForElementNotVisible,
  waitForText: emitWaitForText,
  webdriverAnswerOnVisiblePrompt: emitAnswerOnNextPrompt,
  webdriverChooseCancelOnVisibleConfirmation: emitChooseCancelOnNextConfirmation,
  webdriverChooseCancelOnVisiblePrompt: emitChooseCancelOnNextConfirmation,
  webdriverChooseOkOnVisibleConfirmation: emitChooseOkOnNextConfirmation,
  while: emitControlFlowWhile
};

exporter.register.preprocessors(emitters);

function register(command, emitter) {
  exporter.register.emitter({command, emitter, emitters});
}

function emit(command) {
  return exporter.emit.command(command, emitters[command.command], {
    variableLookup,
    emitNewWindowHandling
  });
}

function canEmit(commandName) {
  return !!emitters[commandName];
}

function variableLookup(varName) {
  return `vars.get("${varName}").toString()`;
}

function variableSetter(varName, value) {
  return varName ? `vars.put("${varName}", ${value});` : '';
}

function emitWaitForWindow() {
  const generateMethodDeclaration = name => {
    return `public String ${name}(int timeout) {`;
  };
  const commands = [
    {level: 0, statement: 'sleep(timeout);'},
    {level: 0, statement: 'Set<String> whNow = WebDriverRunner.getWebDriver().getWindowHandles();'},
    {
      level: 0,
      statement:
        'Set<String> whThen = (Set<String>) vars.get("window_handles");'
    },
    {level: 0, statement: 'if (whNow.size() > whThen.size()) {'},
    {level: 1, statement: 'whNow.removeAll(whThen);'},
    {level: 0, statement: '}'},
    {level: 0, statement: 'return whNow.iterator().next();'}
  ];
  return Promise.resolve({
    name: 'waitForWindow',
    commands,
    generateMethodDeclaration
  });
}

async function emitNewWindowHandling(command, emittedCommand) {
  return Promise.resolve(
    `vars.put("window_handles", WebDriverRunner.getWebDriver().getWindowHandles());\n${await emittedCommand}\nvars.put("${
      command.windowHandleName
    }", waitForWindow(${command.windowTimeout}));`
  );
}

function emitAssert(varName, value) {
  return Promise.resolve(
    `assertEquals(vars.get("${varName}").toString(), "${value}");`
  );
}

function emitAssertAlert(alertText) {
  return Promise.resolve(
    `assertEquals(switchTo().alert().getText(), "${alertText}");`
  );
}

function emitAnswerOnNextPrompt(textToSend) {
  return Promise.resolve(
    `prompt("${textToSend}");`
  );
}

async function emitCheck(locator) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).setSelected(true);`
  );
}

function emitChooseCancelOnNextConfirmation() {
  return Promise.resolve(`dismiss();`);
}

function emitChooseOkOnNextConfirmation() {
  return Promise.resolve(`confirm();`);
}

async function emitClick(target) {
  return Promise.resolve(
    `$(${await location.emit(target)}).click();`
  );
}

async function emitClose() {
  return Promise.resolve(`WebDriverRunner.closeWindow();`);
}

function generateExpressionScript(script) {
  const scriptString = script.script.replace(/"/g, "'");
  return `(Boolean) executeJavaScript("return (${scriptString})"${generateScriptArguments(
    script
  )})`;
}

function emitControlFlowDo() {
  return Promise.resolve({
    commands: [{level: 0, statement: 'do {'}],
    endingLevelAdjustment: 1
  });
}

function emitControlFlowElse() {
  return Promise.resolve({
    commands: [{level: 0, statement: '} else {'}],
    startingLevelAdjustment: -1,
    endingLevelAdjustment: +1
  });
}

function emitControlFlowElseIf(script) {
  return Promise.resolve({
    commands: [
      {
        level: 0,
        statement: `} else if (${generateExpressionScript(script)}) {`
      }
    ],
    startingLevelAdjustment: -1,
    endingLevelAdjustment: +1
  });
}

function emitControlFlowEnd() {
  return Promise.resolve({
    commands: [{level: 0, statement: `}`}],
    startingLevelAdjustment: -1
  });
}

function emitControlFlowIf(script) {
  return Promise.resolve({
    commands: [
      {level: 0, statement: `if (${generateExpressionScript(script)}) {`}
    ],
    endingLevelAdjustment: 1
  });
}

function emitControlFlowForEach(collectionVarName, iteratorVarName) {
  const collectionName = exporter.parsers.capitalize(collectionVarName);
  const iteratorName = exporter.parsers.capitalize(iteratorVarName);
  return Promise.resolve({
    commands: [
      {
        level: 0,
        statement: `ArrayList collection${collectionName} = (ArrayList) vars.get("${collectionVarName}");`
      },
      {
        level: 0,
        statement: `for (int i${iteratorName} = 0; i${iteratorName} < collection${collectionName}.size() - 1; i${iteratorName}++) {`
      },
      {
        level: 1,
        statement: `vars.put("${iteratorVarName}", collection${collectionName}.get(i${iteratorName}));`
      }
    ]
  });
}

function emitControlFlowRepeatIf(script) {
  return Promise.resolve({
    commands: [
      {level: 0, statement: `} while (${generateExpressionScript(script)});`}
    ],
    startingLevelAdjustment: -1
  });
}

function emitControlFlowTimes(target) {
  const commands = [
    {level: 0, statement: `Integer times = ${target};`},
    {level: 0, statement: 'for(int i = 0; i < times; i++) {'}
  ];
  return Promise.resolve({commands, endingLevelAdjustment: 1});
}

function emitControlFlowWhile(script) {
  return Promise.resolve({
    commands: [
      {level: 0, statement: `while (${generateExpressionScript(script)}) {`}
    ],
    endingLevelAdjustment: 1
  });
}

async function emitDoubleClick(target) {
  return Promise.resolve(
    `$(${await location.emit(target)}).doubleClick();`
  );
}

async function emitDragAndDrop(dragged, dropped) {
  return Promise.resolve(
    `$(${await location.emit(dragged)}).dragAndDropTo(${await location.emit(dropped)});`
  );
}

async function emitEcho(message) {
  const _message = message.startsWith('vars.get') ? message : `"${message}"`;
  return Promise.resolve(`System.out.println(${_message});`);
}

async function emitEditContent(locator, content) {
  const commands = [
    {level: 0, statement: '{'},
    {
      level: 1,
      statement: `SelenideElement element = $(${await location.emit(locator)});`
    },
    {
      level: 1,
      statement: `executeJavaScript("if(arguments[0].contentEditable === 'true') {arguments[0].innerText = '${content}'}", element);`
    },
    {level: 0, statement: '}'}
  ];
  return Promise.resolve({commands});
}

async function emitExecuteScript(script, varName) {
  const result = `executeJavaScript("${script.script}"${generateScriptArguments(
    script
  )})`;
  return Promise.resolve(variableSetter(varName, result));
}

async function emitExecuteAsyncScript(script, varName) {
  const result = `executeAsyncJavaScript("var callback = arguments[arguments.length - 1];${
    script.script
  }.then(callback).catch(callback);${generateScriptArguments(script)}")`;
  return Promise.resolve(variableSetter(varName, result));
}

function generateScriptArguments(script) {
  return `${script.argv.length ? ', ' : ''}${script.argv
    .map(varName => `vars.get("${varName}")`)
    .join(',')}`;
}

async function emitMouseDown(locator) {
  return Promise.resolve(
    `actions().moveToElement($(${await location.emit(locator)})).clickAndHold().perform();`
  );
}

async function emitMouseMove(locator) {
  return Promise.resolve(
    `actions().moveToElement($(${await location.emit(locator)})).perform();`
  );
}

async function emitMouseOut() {
  return Promise.resolve(
    `actions().moveToElement($(byTagName("body")), 0, 0).perform();`
  );
}

async function emitMouseUp(locator) {
  return Promise.resolve(
    `actions().moveToElement($(${await location.emit(locator)})).release().perform();`
  );
}

function emitOpen(target) {
  const url = /^(file|http|https):\/\//.test(target)
    ? `"${target}"`
    : `"${global.baseUrl}${target}"`;
  return Promise.resolve(`open(${url});`);
}

async function emitPause(time) {
  return Promise.resolve(`sleep(${time});`);
}

async function emitRun(testName) {
  return Promise.resolve(`${exporter.parsers.sanitizeName(testName)}();`);
}

async function emitRunScript(script) {
  return Promise.resolve(
    `executeJavaScript("${script.script}${generateScriptArguments(script)}");`
  );
}

async function emitSetWindowSize(size) {
  const [width, height] = size.split('x');
  return Promise.resolve(
    `WebDriverRunner.getWebDriver().manage().window().setSize(new Dimension(${width}, ${height}));`
  );
}

async function emitSelect(selectElement, option) {
  return Promise.resolve(
    `$(${await location.emit(selectElement)}).${await selection.emit(option)}`
  );
}

async function emitSelectFrame(frameLocation) {
  if (frameLocation === 'relative=top' || frameLocation === 'relative=parent') {
    return Promise.resolve('switchTo().defaultContent();');
  } else if (/^index=/.test(frameLocation)) {
    return Promise.resolve(
      `switchTo().frame(${Math.floor(
        frameLocation.split('index=')[1]
      )});`
    );
  } else {
    return Promise.resolve(
      `switchTo().frame($(${await location.emit(frameLocation)}));`
    );
  }
}

async function emitSelectWindow(windowLocation) {
  if (/^handle=/.test(windowLocation)) {
    return Promise.resolve(
      `switchTo().window(${windowLocation.split('handle=')[1]});`
    );
  } else if (/^name=/.test(windowLocation)) {
    return Promise.resolve(
      `switchTo().window("${windowLocation.split('name=')[1]}");`
    );
  } else if (/^win_ser_/.test(windowLocation)) {
    if (windowLocation === 'win_ser_local') {
      return Promise.resolve({
        commands: [{
          level: 0,
          statement: 'switchTo().window(0);'
        }]
      });
    } else {
      const index = parseInt(windowLocation.substr('win_ser_'.length));
      return Promise.resolve({
        commands: [{
          level: 0,
          statement: `switchTo().window(${index});`
        }]
      });
    }
  } else {
    return Promise.reject(
      new Error('Can only emit `select window` using handles')
    );
  }
}

function generateSendKeysInput(value) {
  if (typeof value === 'object') {
    return value
      .map(s => {
        if (s.startsWith('vars.get')) {
          return `val(${s})`;
        } else if (s.startsWith('Key[')) {
          const key = s.match(/\['(.*)'\]/)[1];
          return `sendKeys(Keys.${key})`;
        } else {
          return `val("${s}")`;
        }
      })
      .join('.');
  } else {
    if (value.startsWith('vars.get')) {
      return `val(${value})`;
    } else {
      return `val("${value}")`;
    }
  }
}

async function emitSendKeys(target, value) {
  return Promise.resolve(
    `$(${await location.emit(target)}).${generateSendKeysInput(value)};`
  );
}

function emitSetSpeed() {
  return Promise.resolve(
    `System.out.println("\`set speed\` is a no-op in code export, use \`pause\` instead");`
  );
}

async function emitStore(value, varName) {
  return Promise.resolve(variableSetter(varName, `"${value}"`));
}

async function emitStoreAttribute(locator, varName) {
  const attributePos = locator.lastIndexOf('@');
  const elementLocator = locator.slice(0, attributePos);
  const attributeName = locator.slice(attributePos + 1);
  const commands = [
    {level: 0, statement: '{'},
    {
      level: 1,
      statement: `String attribute = $(${await location.emit(elementLocator)}).getAttribute("${attributeName}");`
    },
    {level: 1, statement: `${variableSetter(varName, 'attribute')}`},
    {level: 0, statement: '}'}
  ];
  return Promise.resolve({commands});
}

async function emitStoreText(locator, varName) {
  const result = `$(${await location.emit(locator)}).text()`;
  return Promise.resolve(variableSetter(varName, result));
}

async function emitStoreTitle(_, varName) {
  return Promise.resolve(variableSetter(varName, 'title()'));
}

async function emitStoreValue(locator, varName) {
  const result = `$(${await location.emit(locator)}).getAttribute("value")`;
  return Promise.resolve(variableSetter(varName, result));
}

async function emitStoreWindowHandle(varName) {
  return Promise.resolve(variableSetter(varName, 'WebDriverRunner.getWebDriver().getWindowHandle()'));
}

async function emitStoreXpathCount(locator, varName) {
  const result = `$$(${await location.emit(locator)}).size()`;
  return Promise.resolve(variableSetter(varName, result));
}

async function emitSubmit(locator) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).submit();`
  );
}

async function emitType(target, value) {
  return Promise.resolve(
    `$(${await location.emit(target)}).${generateSendKeysInput(value)};`
  );
}

async function emitUncheck(locator) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).setSelected(false);`
  );
}

async function emitVerifyChecked(locator) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).shouldBe(selected);`
  );
}

async function emitVerifyEditable(locator) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).shouldBe(enabled, not(readonly));`
  );
}

async function emitVerifyElementPresent(locator) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).should(exist);`
  );
}

async function emitVerifyElementNotPresent(locator) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).shouldNot(exist);`
  );
}

async function emitVerifyNotChecked(locator) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).shouldNotBe(checked);`
  );
}

async function emitVerifyNotEditable(locator) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).shouldNotBe(enabled);`
  );
}

async function emitVerifyNotSelectedValue(locator, expectedValue) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).shouldNotHave(value("${exporter.emit.text(expectedValue)}"));`
  );
}

async function emitVerifyNotText(locator, text) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).shouldNotHave(text("${exporter.emit.text(text)}"));`
  );
}

async function emitVerifySelectedLabel(locator, labelValue) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).getSelectedOption().shouldHave(text("${labelValue}"));`
  );
}

async function emitVerifySelectedValue(locator, value) {
  return emitVerifyValue(locator, value);
}

async function emitVerifyText(locator, text) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).shouldHave(text("${exporter.emit.text(text)}"));`
  );
}

async function emitVerifyValue(locator, value) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).shouldHave(value("${value}"));`
  );
}

async function emitVerifyTitle(title) {
  return Promise.resolve(`assertEquals(title(), "${title}");`);
}

async function emitWaitForElementEditable(locator, timeout) {
  const commands = [
    {level: 0, statement: '{'},
    {
      level: 1,
      statement: `$(${await location.emit(
        locator
      )}).waitUntil(visible, ${timeout});`
    },
    {
      level: 1,
      statement: `$(${await location.emit(
        locator
      )}).shouldBe(enabled);`
    },
    {level: 0, statement: '}'}
  ];
  return Promise.resolve({commands});
}

async function emitWaitForText(locator, text) {
  const timeout = 30000;
  return Promise.resolve(
    `$(${await location.emit(locator)}).waitUntil(text("${text}"), ${timeout});`
  );
}

function skip() {
  return Promise.resolve('');
}

async function emitWaitForElementPresent(locator, timeout) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).waitUntil(exist, ${timeout});`
  );
}

async function emitWaitForElementVisible(locator, timeout) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).waitUntil(visible, ${timeout});`
  );
}

async function emitWaitForElementNotEditable(locator, timeout) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).waitUntil(not(enabled), ${timeout});`
  );
}

async function emitWaitForElementNotPresent(locator, timeout) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).waitUntil(not(exist), ${timeout});`
  );
}

async function emitWaitForElementNotVisible(locator, timeout) {
  return Promise.resolve(
    `$(${await location.emit(locator)}).waitUntil(disappears, ${timeout});`
  );
}

export default {
  canEmit,
  emit,
  register,
  extras: {emitWaitForWindow}
};
