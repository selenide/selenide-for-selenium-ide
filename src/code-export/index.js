import {codeExport as exporter} from '@seleniumhq/side-utils';
import emitter from './command';
import location from './location';
import {generateHooks} from './hook';

// Define language options
export const displayName = 'Java Selenide';

export let opts = {};
opts.emitter = emitter;
opts.hooks = generateHooks();
opts.fileExtension = '.java';
opts.commandPrefixPadding = '  ';
opts.terminatingKeyword = '}';
opts.commentPrefix = '//';
opts.generateMethodDeclaration = generateMethodDeclaration;

// Create generators for dynamic string creation of primary entities (e.g., filename, methods, test, and suite)
function generateTestDeclaration(name) {
  return `@Test\npublic void ${exporter.parsers.uncapitalize(
    exporter.parsers.sanitizeName(name)
  )}() {`;
}

function generateMethodDeclaration(name) {
  return `public void ${exporter.parsers.sanitizeName(name)}() {`;
}

function generateSuiteDeclaration(name) {
  return `public class ${exporter.parsers.capitalize(
    exporter.parsers.sanitizeName(name)
  )}Test {`;
}

function generateFilename(name) {
  return `${exporter.parsers.capitalize(
    exporter.parsers.sanitizeName(name)
  )}Test${opts.fileExtension}`;
}

// Emit an individual test, wrapped in a suite (using the test name as the suite name)
export async function emitTest({
  baseUrl,
  test,
  tests,
  project,
  enableOriginTracing,
  beforeEachOptions,
  enableDescriptionAsComment
}) {
  global.baseUrl = baseUrl;
  const testDeclaration = generateTestDeclaration(test.name);
  const result = await exporter.emit.test(test, tests, {
    ...opts,
    testDeclaration,
    enableOriginTracing,
    enableDescriptionAsComment,
    project
  });
  const suiteName = test.name;
  const suiteDeclaration = generateSuiteDeclaration(suiteName);
  const _suite = await exporter.emit.suite(result, tests, {
    ...opts,
    suiteDeclaration,
    suiteName,
    project,
    beforeEachOptions
  });
  return {
    filename: generateFilename(test.name),
    body: exporter.emit.orderedSuite(_suite)
  };
}

// Emit a suite with all of its tests
export async function emitSuite({
  baseUrl,
  suite,
  tests,
  project,
  enableOriginTracing,
  beforeEachOptions,
  enableDescriptionAsComment
}) {
  global.baseUrl = baseUrl;
  const result = await exporter.emit.testsFromSuite(tests, suite, opts, {
    enableOriginTracing,
    enableDescriptionAsComment,
    generateTestDeclaration,
    project
  });
  const suiteDeclaration = generateSuiteDeclaration(suite.name);
  const _suite = await exporter.emit.suite(result, tests, {
    ...opts,
    suiteDeclaration,
    suite,
    project,
    beforeEachOptions
  });
  return {
    filename: generateFilename(suite.name),
    body: exporter.emit.orderedSuite(_suite)
  };
}

export default {
  emit: {
    test: emitTest,
    suite: emitSuite,
    locator: location.emit
  },
  register: {
    command: emitter.register,
    dependency: opts.hooks.declareDependencies.register,
    inEachBegin: opts.hooks.inEachBegin.register,
    inEachEnd: opts.hooks.inEachEnd.register
  }
};
