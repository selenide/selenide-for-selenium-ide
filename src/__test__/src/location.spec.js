import {emit} from '../../code-export/location'

describe('location code emitter', () => {
  it('should fail to emit empty string', () => {
    return expect(() => {
      emit('')
    }).toThrow("Locator can't be empty")
  })
  it('should fail to emit unknown locator', () => {
    return expect(() => {
      emit('notExists=element')
    }).toThrow('Unknown locator notExists')
  })
  it('should emit id locator', () => {
    const type = 'id'
    const selector = 'someId'
    return expect(emit(`${type}=${selector}`)).resolves.toBe(
      `Selectors.byId("${selector}")`
    )
  })
  it('should emit link locator', () => {
    const type = 'link'
    const selector = 'someLink'
    return expect(emit(`${type}=${selector}`)).resolves.toBe(
      `Selectors.byLinkText("${selector}")`
    )
  })
  it('should emit linkText locator', () => {
    const type = 'linkText'
    const selector = 'someLink'
    return expect(emit(`${type}=${selector}`)).resolves.toBe(
      `Selectors.byLinkText("${selector}")`
    )
  })
  it('should emit partialLinkText locator', () => {
    const type = 'partialLinkText'
    const selector = 'someLink'
    return expect(emit(`${type}=${selector}`)).resolves.toBe(
      `Selectors.byPartialLinkText("${selector}")`
    )
  })
  it('should emit css locator', () => {
    const type = 'css'
    const selector = 'someCss'
    return expect(emit(`${type}=${selector}`)).resolves.toBe(
      `"${selector}"`
    )
  })
  it('should emit css locator with `=` sign', () => {
    const type = 'css'
    const selector = 'a[title=JScript]'
    return expect(emit(`${type}=${selector}`)).resolves.toBe(
      `"${selector}"`
    )
  })
  it('should escape quotes in locator strings', () => {
    const type = 'css'
    const selector = 'a[title="escaped"]'
    return expect(emit(`${type}=${selector}`)).resolves.toMatchSnapshot()
  })
  it('should emit xpath locator', () => {
    const type = 'xpath'
    const selector = 'someXpath'
    return expect(emit(`${type}=${selector}`)).resolves.toBe(
      `Selectors.byXpath("${selector}")`
    )
  })
  it('should emit implicit xpath locator', () => {
    const selector = '//test=xpath'
    return expect(emit(selector)).resolves.toBe(`Selectors.byXpath("${selector}")`)
  })
  it('should emit name locator', () => {
    const type = 'name'
    const selector = 'someName'
    return expect(emit(`${type}=${selector}`)).resolves.toBe(
      `Selectors.byName("${selector}")`
    )
  })
})