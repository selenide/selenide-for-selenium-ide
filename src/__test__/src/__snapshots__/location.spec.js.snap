// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`location code emitter should escape quotes in locator strings 1`] = `"\\"a[title=\\"escaped\\"]\\""`;
